streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/StickyTaskAssignor.java

```diff
@@ -0,0 +1,145 @@
+        this.clients = clients;
+        this.taskIds = taskIds;
+        mapPreviousTaskAssignment(clients);
+    }
+
+    @Override
+    public void assign(final int numStandbyReplicas) {
+        assignActive();
+        assignStandby(numStandbyReplicas);
+    }
+
+    // Visible for testing
+    void assignStandby(final int numStandbyReplicas) {
+        for (int i = 0; i < numStandbyReplicas; i++) {
+            for (final TaskId taskId : taskIds) {
+                final Set<ID> ids = findClientsWithoutAssignedTask(taskId);
+                if (ids.isEmpty()) {
+                    continue;
+                }
+                final ClientState<TaskId> client = findClient(taskId, ids);
+                client.assign(taskId, false);
+            }
+        }
+    }
+
+    private void assignActive() {
+        for (final TaskId taskId : taskIds) {
+            final ClientState<TaskId> client = findClient(taskId, clients.keySet());
+            client.assign(taskId, true);
+        }
+    }
+
+    private Set<ID> findClientsWithoutAssignedTask(final TaskId taskId) {
+        final Set<ID> clientIds = new HashSet<>();
+        for (final Map.Entry<ID, ClientState<TaskId>> client : clients.entrySet()) {
+            if (!client.getValue().hasAssignedTask(taskId)) {
+                clientIds.add(client.getKey());
+            }
+        }
+        return clientIds;
+    }
+
+
+    private ClientState<TaskId> findClient(final TaskId taskId, final Set<ID> clientsWithin) {
+        // optimize the case where there is only 1 id to search within.
+        if (clientsWithin.size() == 1) {
+            return clients.get(clientsWithin.iterator().next());
+        }
+
+        final ClientState<TaskId> previous = clientWithPreviousAssignment(taskId, clientsWithin);
+        if (previous == null) {
+            return leastLoaded(clientsWithin);
+        }
+
+        if (previous.reachedCapacity()) {
+            final ClientState<TaskId> standby = findLeastLoadedClientWithStandby(taskId, clientsWithin);
+            if (standby == null || standby.reachedCapacity()) {
+                return leastLoaded(clientsWithin);
+            }
+            return standby;
+        }
+
+        return previous;
+    }
+
+    private ClientState<TaskId> clientWithPreviousAssignment(final TaskId taskId, final Set<ID> clientsWithin) {
+        final ID previous = previousActiveTaskAssignment.get(taskId);
+        if (previous != null && clientsWithin.contains(previous)) {
+            return clients.get(previous);
+        }
+        return findLeastLoadedClientWithStandby(taskId, clientsWithin);
+    }
+
+    private ClientState<TaskId> findLeastLoadedClientWithStandby(final TaskId taskId, final Set<ID> clientsWithin) {
+        final Set<ID> ids = previousStandbyTaskAssignment.get(taskId);
+        if (ids == null) {
+            return null;
+        }
+        final HashSet<ID> constrainTo = new HashSet<>(ids);
+        constrainTo.retainAll(clientsWithin);
```

Comment:
`constrainTo` is never used. I guess you can remove both lines. `ids` should not contain any invalid `ID`s anyway -- otherwise it's a bug and `retainAll` would only mask the bug, but not fix it.